elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/filter/visitor/SplitFilterExpressionVisitor.java

```diff
@@ -0,0 +1,187 @@
+                    left.getWhereExpression(),
+                    right.getHavingExpression()
+            );
+        } else {
+            return WhereHaving.withWhereAndHaving(
+                    AndFilterExpression.and(
+                            left.getWhereExpression(),
+                            right.getWhereExpression()
+                    ),
+                    AndFilterExpression.and(
+                            left.getHavingExpression(),
+                            right.getHavingExpression()
+                    )
+            );
+        }
+    }
+
+    @Override
+    public WhereHaving visitOrExpression(final OrFilterExpression expression) {
+        WhereHaving left = expression.getLeft().accept(this);
+        WhereHaving right = expression.getRight().accept(this);
+
+        if (left.isWhereExpression() && right.isWhereExpression()) {
+            // pure-W OR pure-W
+            return WhereHaving.pureWhere(
+                    OrFilterExpression.or(
+                            left.getWhereExpression(),
+                            right.getWhereExpression()
+                    )
+            );
+        } else {
+            // H OR H
+            // W OR H (W promote)
+            // H OR W (W promote)
+            return WhereHaving.pureHaving(
+                    OrFilterExpression.or(
+                            expression.getLeft(),
+                            expression.getRight()
+                    )
+            );
+        }
+    }
+
+    @Override
+    public WhereHaving visitNotExpression(NotFilterExpression expression) {
+        FilterExpression normalized = getNormalizationVisitor().visitNotExpression(expression);
+
+        if (normalized instanceof AndFilterExpression) {
+            return visitAndExpression((AndFilterExpression) normalized);
+        } else if (normalized instanceof OrFilterExpression) {
+            return visitOrExpression((OrFilterExpression) normalized);
+        } else if (normalized instanceof NotFilterExpression) {
+            normalized = ((NotFilterExpression) normalized).getNegated();
+            return visitNotExpression((NotFilterExpression) normalized);
+        } else {
+            return visitPredicate((FilterPredicate) normalized);
+        }
+    }
+
+    /**
+     * Returns whether or not a {@link FilterPredicate} corresponds to a {@code HAVING} clause in JPQL query.
+     * <p>
+     * A {@link FilterPredicate} corresponds to a {@code HAVING} clause iff the predicate field has
+     * {@link MetricAggregation} annotation on it.
+     *
+     * @param filterPredicate  The terminal filter expression to check for
+     *
+     * @return {@code true} if the {@link FilterPredicate} is a HAVING clause
+     */
+    private boolean isHavingPredicate(final FilterPredicate filterPredicate) {
+        Class entityClass = filterPredicate.getEntityType();
+        String fieldName = filterPredicate.getField();
+
+        return getEntityDictionary()
```

Comment:
This is where the schema will come in handy.  MetricComputation might also require Having (not just MetricAggregation).  

Instead of using the EntityDictionary, this could leverage the schema.